zero_nova/zero.asm

```diff
@@ -1,15 +1,72 @@
-machine Zero {
-	degree 2;
+machine NovaZero {
 
-	reg pc[@pc];
-	reg X[<=];
+    degree 10;
 
-	reg A;
+    // this simple machine does not have submachines
 
-	instr ASSERT_IS_ZERO_A { A = 0 }
+    reg pc[@pc];
+    reg X[<=];
+    reg Y[<=];
+    reg Z[<=];
+    reg x0;
+    reg x1;
 
-	function main {
-		A <=X= ${ ("input", 0) }; // pc = 0 => input[0] = X && X = A
-		ASSERT_IS_ZERO_A;         // pc = 1 => A = 0
-	}
+    constraints {
+        col witness x_b0; // TODO support to 64
+        col witness x_b1;
+        col witness x_b2;
+        col witness x_b3;
+
+        // constraints bit
+        x_b0 * (1-x_b0) = 0;
+        x_b1 * (1-x_b1) = 0;
+        x_b2 * (1-x_b2) = 0;
+        x_b3 * (1-x_b3) = 0;
+        // ...
+    }
+
+    instr incr X -> Y {
+        // range check on X
+        // Y = Z + 1, // this one work means assignment register signature `X -> Y` doen't take effect lol ?
```

Comment:
yes I aware of this. 

As the time of writing, now each instruction augmented circuit only parse the constraints inside the instruction definition, but not parsing the general constraints. For efficiency concern, I haven't parse general constraints in each augmented circuit

@leonardoalt I think there are 2 design choices
1. `inline` support: since each instruction is self-contained in current folding schems, do you think it make sense to have some `inline` feature during asm -> pil compiler, so all the constraints being touch by instruction A, it will be layout in the line of i.e. `instr_A * (all the constrains chain)` ? 
2. `nova augment circuit builder optimisation`: introduce some lazily constraints compiling strategy in nova augment circuit parsing, such as DFS recursive traverse to expand polynomial reference in constraints area. This is optimise so other un-related constraints will not be compiled into the circuit

Reply:
1. Currently this is done like this, but I think inside constraints for next-registers. Need to think about it.
2. This direction sounds good, yes.

Another thing that I haven't understood so far is, what happens to the `constraints` block of a machine? These constraints should be true for every line, so are they used in the single-instruction-circuits?